doc/stdenv: rewrite manual build procedure to be closer to an auto-build (#262137)
NixOS/nixpkgs: doc/stdenv/stdenv.chapter.md

```bash
# Go to an empty directory
cd "$(mktemp -d)"

nix-shell '<nixpkgs>' -A some_package
# Or, equivalently, using the experimental nix-command feature:
SHELL=bash nix --extra-experimental-features nix-command develop -f '<nixpkgs>' some_package

# Set the output variables used by the build to somewhere writable:
export out=$(pwd)/out

# Unpack sources and enter source root:
phases="${prePhases[*]:-} unpackPhase patchPhase" genericBuild

# Run some phases until you reach the failure; for example, if the failure is
# in the build phase:
phases="${preConfigurePhases[*]:-} configurePhase ${preBuildPhases[*]:-} buildPhase" genericBuild

# Then rerun the problem phase as many times as desired, e.g.:
phases="buildPhase" genericBuild
```

To modify a [phase](#sec-stdenv-phases), first print it with

```bash
echo "$buildPhase"
# Or, if that is empty, for instance, if it is using a function:
type buildPhase
```

then change it in a text editor, and paste it back to the terminal.

> Note that this method may have some inconsistencies in environment variables compared to a normal build within the sandbox.

Author: added

### Building a `stdenv` package in `nix-shell` {#sec-building-stdenv-package-in-nix-shell}

To build a `stdenv` package in a [`nix-shell`](https://nixos.org/manual/nix/unstable/command-ref/nix-shell.html), enter a shell, find the [phases](#sec-stdenv-phases) you wish to build, then invoke `genericBuild` manually:

Go to an empty directory, invoke `nix-shell` with the desired package, and from inside the shell, set the output variables to a writable directory:

```bash
cd "$(mktemp -d)"
nix-shell '<nixpkgs>' -A some_package
export out=$(pwd)/out
```

Next, invoke the desired parts of the build.
First, run the phases that generate a working copy of the sources, which will change directory to the sources for you:

```bash
phases="${prePhases[*]:-} unpackPhase patchPhase" genericBuild
```

Then, run more phases up until the failure is reached.
For example, if the failure is in the build phase, the following phases would be required:

```bash
phases="${preConfigurePhases[*]:-} configurePhase ${preBuildPhases[*]:-} buildPhase" genericBuild
```

> Rerun a single phase as many times as necessary to examine the failure like so:

Reviewer: ```suggestion
Re-run a single phase as many times as necessary to examine the failure like so:
```

```bash
phases="buildPhase" genericBuild
```

To modify a [phase](#sec-stdenv-phases), first print it with

```bash
echo "$buildPhase"
```

Or, if that is empty, for instance, if it is using a function:

```bash
type buildPhase
```

then change it in a text editor, and paste it back to the terminal.

::: {.note}
This method may have some inconsistencies in environment variables and behaviour compared to a normal build within the [Nix build sandbox](https://nixos.org/manual/nix/unstable/language/derivations#builder-execution).
The following is a non-exhaustive list of such differences:

- `TMP`, `TMPDIR`, and similar variables likely point to non-empty directories that the build might conflict with files in.
- Output store paths are not writable, so the variables for outputs need to be overridden to writable paths.
- Other environment variables may be inconsistent with a `nix-build` either due to `nix-shell`'s initialization script or due to the use of `nix-shell` without the `--pure` option.

> If the build fails differently inside the shell than in the sandbox, consider using [breakpointHook](#breakpointhook) and invoking a `nix-build` instead.

Reviewer: ```suggestion
If the build fails differently inside the shell than in the sandbox, consider using [`breakpointHook`](#breakpointhook) and invoking `nix-build` instead.
```

> The `--keep-failed` option for `nix-build` may also be useful to examine the build directory of a failed build.

Reviewer: ```suggestion
The [`--keep-failed`](https://nixos.org/manual/nix/unstable/command-ref/conf-file#opt--keep-failed) option for `nix-build` may also be useful to examine the build directory of a failed build.
```
